src/persist-client/src/internal/encoding.rs

```diff
@@ -1607,4 +1614,48 @@ mod tests {
 
         proptest!(|(state in any_state::<u64>(0..3))| testcase(state));
     }
+
+    #[mz_ore::test]
+    fn check_data_versions() {
+        #[track_caller]
+        fn testcase(code: &str, data: &str, expected: Result<(), ()>) {
+            let code = Version::parse(code).unwrap();
+            let data = Version::parse(data).unwrap();
+            #[allow(clippy::disallowed_methods)]
+            let actual =
+                std::panic::catch_unwind(|| check_data_version(&code, &data)).map_err(|_| ());
+            assert_eq!(actual, expected);
+        }
+
+        testcase("0.10.0-dev", "0.10.0-dev", Ok(()));
+        testcase("0.10.0-dev", "0.10.0", Ok(()));
+        testcase("0.10.0-dev", "0.11.0-dev", Ok(()));
+        testcase("0.10.0-dev", "0.11.0", Ok(()));
```

Comment:
I find this one (`0.10.0-dev -> 0.11.0`, and similarly `0.10.0-dev -> 0.11.0-dev`) an interesting policy question! Should we force a dev version to upgrade to the non-dev before moving on? I don't think this is super common, and disallowing it probably makes things marginally easier to reason about, so maybe yes?

Reply:
I'm not sure what the right policy is, but I feel like this is _probably_ not the right place to enforce it... what you have here seems good to me!

Reply:
Ideally we never see a `-dev` version in prod so whatever we decide won't have a huge impact. 

I can see `0.X.Y -> 0.X+1.0-dev` being very useful for the upgrade checker tool to check can some env upgrade to `main`. 
I can also see `0.10.0-dev -> 0.11.0-dev` maybe being useful for tests where we upgrade between two dev versions, though I don't know why we'd want to test that. 
`0.10.0-dev -> 0.11.0` might be useful for people with custom builds on their staging envs that want to go back to the normal releases.

Reply:
Thinking, like, deploy tooling? 

Concretely: I can easily end up with dev builds that are mutually incompatible with the "real" branch. (eg. adding a new field in a branch, deploying it to staging, then deciding I want to serialize it in a different incompatible way and merging the second version.) There's not really a guarantee of even backwards compat for dev builds... but obviously it's a very convenient thing to assume when we roll out staging every week! So my gut is that if we decide we want to enforce this sort of thing, the sort of deploy tool that understands the difference between prod and staging (and might eg. block deploying dev builds to prod) seems like a fine choice.

Reply:
> I can see `0.X.Y -> 0.X+1.0-dev` being very useful for the upgrade checker tool to check can some env upgrade to `main`.

Ah, yeah, that's a great observation. I'll add that as context

> I can also see `0.10.0-dev -> 0.11.0-dev` maybe being useful for tests where we upgrade between two dev versions, though I don't know why we'd want to test that.

Might be more convenience (two random shas on main) than intentionally, also a good point.

> `0.10.0-dev -> 0.11.0` might be useful for people with custom builds on their staging envs that want to go back to the normal releases.

I'd be okay making this situation require going through `0.10.0` if it made persist's lives easier, but I think the above makes this moot.

> Thinking, like, deploy tooling?
> 
> Concretely: I can easily end up with dev builds that are mutually incompatible with the "real" branch. (eg. adding a new field in a branch, deploying it to staging, then deciding I want to serialize it in a different incompatible way and merging the second version.) There's not really a guarantee of even backwards compat for dev builds... but obviously it's a very convenient thing to assume when we roll out staging every week! So my gut is that if we decide we want to enforce this sort of thing, the sort of deploy tool that understands the difference between prod and staging (and might eg. block deploying dev builds to prod) seems like a fine choice.

Overall, great point. We don't run -dev in prod, and I'm overthinking things here. If something truly wonky happens, it'll be in staging.

We'll definitely have to figure out our best practices here as we go. At 4sq, we ended up with a policy of treating anything that had merged to main as if it had gone to prod (i.e. if you decide to serialize in a different way as in your example, always gotta do the migration). It's more work, but the way mz uses protos (immediately deserializing them into a non-proto) is modeled after where we ended up at 4sq and designed to make that as easy as possible plus contain the tech debt. 4sq tried originally a policy of being able to make incompatible changes if they hadn't rolled out yet, but it was surprisingly tricky to keep correct in practice. Otoh, 4sq had a bunch of microservices that were all deployed on separate schedules whereas mz has a weekly release of everything, which might completely change the math on this.